### PR TITLE
Add Springer Nature engine

### PR DESCRIPTION
**Pull request by pchg**
## What does this PR do?

Add a search engine for Springer Nature: https://www.springernature.com/ .

## Why is this change important?

Springer Nature is a major player in the scientific community, its publications deserve to be searched.

## How to test this PR locally?

First get your API key following this page:
https://dev.springernature.com/signup

In searx/engines/springer.py
at line 21, add this API key.
I left my own key, commented out on line 20.  Feel free to use it, if needed.

In searx, try a search using some scientific terms, links to Springer Nature should be listed.
Here is an example of a screenshot from such a search:
image

## Related issues

None

### CODE REVIEW

**Comment by return42** [comment]
FYI: You should use `make test` before commit, have a look at our [Development Quickstart](http://searx.github.io/searx/dev/quickstart.html) and [How to contribute](http://searx.github.io/searx/dev/contribution_guide.html).

File: searx/settings.yml [hunk 1]
```diff
@@ -953,6 +953,11 @@ engines:
 #    query_fields : '' # query fields
 #    enable_http : True
 
+  - name : springer nature
+    engine : springer
+    shortcut : springer
+    categories : science
+
```
Comment (return42) [inline searx/settings.yml:967]: I tested some queries of the springer database (e.g. [physics](https://api.springernature.com/metadata/json?&q=physics&api_key=a69685087d07eca9f13db62f65b8f601)), for large results I would recommend a timeout of 6 sec

        timeout : 6

**Comment by return42** [comment]
Thanks a lot for this engine which queries a database from very good quality!

I think we should add paging support, the API has the needed [querystring-parameters](https://dev.springernature.com/querystring-parameters) `p` and `s` we need for.  The response includes much more fields besides title and url we can use

Are you able to add paging support and have a look which fields we can use also? .. otherwise I can prepare a patch for you to fetch.

Again thanks, I really like this database.

File: searx/settings.yml (hunk 1 above)
Comment (pchg) [reply searx/settings.yml:967]: Done.

**Comment by pchg** [comment]
> Thanks a lot for this engine which queries a database from very good quality!

Glad you enjoy!


> I think we should add paging support, the API has the needed [querystring-parameters](https://dev.springernature.com/querystring-parameters) `p` and `s` we need for. The response includes much more fields besides title and url we can use
> 
> Are you able to add paging support and have a look which fields we can use also? .. otherwise I can prepare a patch for you to fetch.

Oops.  I'm afraid that I don't really feel able to do so...  I've battled quite a bit to manage to have the results with title and URL, I was so happy with that that I just stopped touching anything!

> Again thanks, I really like this database.

You're welcome!  I like it too, as a scientist.

**Comment by pchg** [comment]
> Are you able to add paging support and have a look which fields we can use also? .. otherwise I can prepare a patch for you to fetch.

I'm trying to implement this, with the help of some knowledgeable friends.  It sorts of works, a bit.

**Comment by pchg** [comment]
> > Are you able to add paging support and have a look which fields we can use also? .. otherwise I can prepare a patch for you to fetch.
> I'm trying to implement this, with the help of some knowledgeable friends. It sorts of works, a bit.

I've tried to do that paging support, but unfortunately, the commit I made does not appear to be related to the branch.  I'm sorry about that, I just don't understand how this could have happened.

I've had some trouble making this pull request: I've tried to git commit --amend several times, and then to git push -f, but every time, it generated conflicts for unknown reasons.  So I rather chose the solution to redo a branch from scratch, and to make several git commits.
Maybe this is what caused the mess...

Sorry about all that mess, I'm not too familiar with the git workflow.

**Comment by return42** [comment]
> I've tried to do that paging support, ... So I rather chose the solution to redo a branch from scratch

If you want, I can do it for you.  But I don't have time for ATM.. maybe at the weekend .. if you have time, lets wait and do not reopen another PR.

**Comment by pchg** [comment]
> > I've tried to do that paging support, ... So I rather chose the solution to redo a branch from scratch
> 
> If you want, I can do it for you. But I don't have time for ATM.. maybe at the weekend .. if you have time, lets wait and do not reopen another PR.

That would be very kind of you, thanks.  No problem, take your time!

**Comment by return42** [comment]
I prepared https://github.com/return42/searx/commits/springer_nature branch with commit b6b07897 and 3fbeec62

Do you know how to force pull from my branch and force push your to your remote branch?   Otherwise I can write down some instructions for you.

---

I added some more fields (publisher, DOI) and the abstract ..

image

**Comment by pchg** [comment]
> I prepared https://github.com/return42/searx/commits/springer_nature branch with commit [b6b0789](https://github.com/searx/searx/commit/b6b078975fa228d3d3542836b70ad1a270f00013) and [3fbeec6](https://github.com/searx/searx/commit/3fbeec627929f523ebae3932c68abda4936e5765)
> 
> Do you know how to force pull from my branch and force push your to your remote branch? Otherwise I can write down some instructions for you.

I suppose that I should add a remote pointing to your repository, and then pull from it, but I'm afraid that the use of "force" tends to frighten me a bit...

So I wouldn't be against some instructions, if you please.

**Comment by pchg** [comment]
> I prepared https://github.com/return42/searx/commits/springer_nature branch with commit [b6b0789](https://github.com/searx/searx/commit/b6b078975fa228d3d3542836b70ad1a270f00013) and [3fbeec6](https://github.com/searx/searx/commit/3fbeec627929f523ebae3932c68abda4936e5765)
> 
> Do you know how to force pull from my branch and force push your to your remote branch? Otherwise I can write down some instructions for you.

I did this differently: I downloaded your code, and merged your modifications and additions into my code by using my good old vimdiff.
Sorry, I feel extremely more comfortable with that way of working than with gitting all over the place.  I will learn...  But now I'm on a bit of a hurry.

I git pushed -f to my repository, so the PR should be updated, hopefully.

When I tested Searx, I didn't get any more results from Springer, though, I'm a bit worried...  I must have done something wrong.

**Comment by return42** [comment]
A short introduction to a git **force pull and force push** workflow.

**Clone** your repo (your github fork of the searx project):

    $ cd ~/Downloads/
    $ git clone https://github.com/pchg/searx.git
    $ cd searx

and checkout local the *feature branch* `origin/springer_nature` (branch of this PR):

    $ git checkout springer_nature
    Switched to branch 'springer_nature'
    Your branch is up to date with 'origin/springer_nature'.

Now we have to **add the remote** from which we want to pull:

    $ git remote add return42 https://github.com/return42/searx.git

    $ git remote -v
    origin      https://github.com/pchg/searx.git (fetch)
    origin      https://github.com/pchg/searx.git (push)
    return42    https://github.com/return42/searx.git (fetch)
    return42    https://github.com/return42/searx.git (push)

Looks good, `origin` is the remote we cloned from and `return42` is the alias name of the foreign remote.  Now lets have a look at the branches we got ..

    $ git branch -av
      master                          11070e57 Merge pull request #2657 from cyclaero/conditional-sigusr1
    * springer_nature                 c6e2e730 Merge branch 'master' into springer_nature
    ...
      remotes/origin/master           11070e57 Merge pull request #2657 from cyclaero/conditional-sigusr1
      remotes/origin/springer_nature  c6e2e730 Merge branch 'master' into springer_nature

OK, now we have two local branches `master` and `springer_nature`.  In the working tree we have checked out the latter one (tagged by an asterisk).  We also have some remote branches from `origin`, but where are *my* `return42` branches?  Oops, we forgot to fetch them from the remote.  To *be up-to-date*, lets fetch from all:

    $ git fetch --all
    Fetching origin
    Fetching return42
    ...
    $ git branch -av
    ...
    remotes/return42/master            e0c903f0 brand: darmarit.org/searx
    ...
    remotes/return42/springer_nature   3fbeec62 [enh] themes - add result.metadata to result sub header

OK, now we have all the branch & patches we need.   Before we start, lets state to see what we have *before* and *after*:

- Working tree is branch `springer_nature` at commit c6e2e730.
- What we want is; that the local branch becomes commit 3fbeec62 from `return42/springer_nature`.

To do this we *force pull* into our local branch, or more simple: lets reset our local branch:

    $ git reset --hard return42/springer_nature

Isn't it cool /  I can't believe, lets have a closer look:

    $ git log --oneline
    3fbeec62 (HEAD -> springer_nature, return42/springer_nature) [enh] themes - add result.metadata to result sub header
    b6b07897 [enh] Add Springer Nature engine
    ae0b621e [mod] Dockerfile: remove protobuff dependency
    ...

Yes our local branch (and working tree) is now at 3fbeec62.  And whats about `origin/springer_nature`

    $ git log --oneline origin/springer_nature
    c6e2e730 (origin/springer_nature) Merge branch 'master' into springer_nature
    ...

OK now we have to **force push** our local branch `springer_nature` to the remote `origin` (your fork of searx):

    $ git push -f origin springer_nature

**Bonus tutorial**: *rebase on upstream*

Our *upstream* repository is the searx project and while we discussed your PR the `master` branch of the upstream progresses.  The lecture now is to put the commits from our PR (branch `origin/springer_nature`) on top of this development (on top of `upstream/master`).   The name for the git workflow is *rebase on upstream*.

For this we need to add the searx project as another remote to our repository and fetch from (all remotes):

    $ git remote add upstream https://github.com/searx/searx.git
    $ git fetch --all

In the example above we added the searx project and named the remote URL `upstream`, to rebase we can use:

    $ git rebase upstream/master

At least we have to push our local branch to the PR branch (`origin`):

    $ git push -f origin springer_nature

The command `rebase` *pulls* from `upstream/master` and tries to merge our *feature* branch (`origin/springer_nature` / the PR) on top, like shown in chapter ["The Rebase Option"](https://www.atlassian.com/git/tutorials/merging-vs-rebasing) (see picture below).

[In case of conflict, git rebase will stop at the first problematic commit and leave conflict markers in the tree. See `git-rebase` DESCRIPTION](https://git-scm.com/docs/git-rebase) how you can `--abort` or `--continue`.

rebase feature branch

To give an overview, here is a abstract image of the collaboration infrastructure we have used:

rebase feature branch

The image is from the [Distributed Workflows](https://git-scm.com/book/en/v2/Distributed-Git-Distributed-Workflows), the only thing I missed is an arrow from *developer (A) public* to *developer (B) private*.

**Comment by mrwormo** [comment]
Whao, thank you so much fir this wonderful explanation @return42 !

**Comment by pchg** [comment]
> A short introduction to a git **force pull and force push** workflow.
=> I like the "short"...

Thanks a lot, I'll do that!

**Comment by pchg** [comment]
Thanks a million @return42 , I followed your very detailed instructions carefully.

I just ran into a small hickup, due to a conflict of several github profiles in my configuration: so instead of doing an initial git clone, I just worked from my usual searx local repository, which is well-configured with the appropriate github account.  And it seems that everything went file!

**Review by return42 — APPROVED** [review]

**Comment by return42** [comment]
>  And it seems that everything went file!

Yes, perfect ;-) .. thanks for taking the time!

> I just ran into a small hickup, 

The author of your commit was https://github.com/pierrechtux while the branch of this PR is from https://github.com/pchg

**Comment by pchg** [comment]
> The author of your commit was https://github.com/pierrechtux while the branch of this PR is from https://github.com/pchg

This is the same person, myself... I'm having trouble managing those accounts, sorry about that mess... At some point, I'll merge those repositories and just keep one single account, to avoid more confusion...

**Comment by return42** [comment]
This engine is ready to merge .. one hint I will give: 

In d5b7d0f78 I added a new field to the them macros **result.metadata** at the bottom of this comment you will find a  is a screenshot how this looks like. **result.metadata**  is the (bold) line below the grayed date.

I added  **result.metadata** because when developing scrappers I often missed a field like this. ATM it is implemented as a simple string, it could be a list or dict  ... what do other think about?

It might be a bit offtopic here but I also miss a **result.doi** field which (e.g.) could be handled by the DOI plugin and rendered as a link (in the metadata).

image

**Review by return42 — APPROVED** [review]

**Comment by return42** [comment]
@pchg instead of _merge from upstream/master_ it is better to _rebase on upstream/master_, see [Bonus totorial :-)](https://github.com/searx/searx/pull/2663#issuecomment-808886980) ... would you like to give it a try?

**Comment by pchg** [comment]
> @pchg instead of _merge from upstream/master_ it is better to _rebase on upstream/master_, see [Bonus totorial :-)](https://github.com/searx/searx/pull/2663#issuecomment-808886980) ... would you like to give it a try?

Actually, all I did was stupidly click on the green button in the github interface... My bad.

I'll do a rebase some time later (am on duty now).

**Comment by return42** [comment]
> Actually, all I did was stupidly click on the green button in the github interface... My bad.

No problem .. this is what github's GUI offers .. I also do miss a rebase button :-)

**Review by return42 — APPROVED** [review]

**Comment by kvch** [comment]
Added in https://github.com/searx/searx/commit/3a0f896b68b24957421c593dc2bfacdd36b8bcb3
Thank you

**Pull request closed without merging**